File: tests/src/com.google.guava/guava/33.6.0-jre/src/test/java/com_google_guava/guava/ResourcesTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_guava.guava;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;

import com.google.common.io.Resources;
import java.net.URL;
import org.junit.jupiter.api.Test;

public class ResourcesTest {
    @Test
    void getResourceFindsClasspathResourceWithContextClassLoader() throws Exception {
        URL resource = Resources.getResource("com_google_guava/guava/resources-classloader.txt");

        assertThat(Resources.toString(resource, UTF_8)).isEqualTo("loaded by classloader\n");
    }

    @Test
    void getResourceFindsRelativeResourceWithContextClass() throws Exception {
        URL resource = Resources.getResource(ResourcesTest.class, "resources-relative.txt");

        assertThat(Resources.toString(resource, UTF_8)).isEqualTo("loaded relative to test class\n");
    }
}

File: tests/src/com.typesafe.play/play_2.13/2.9.10/src/test/scala/com_typesafe_play/play_2_13/ResourcesTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_typesafe_play.play_2_13

import java.net.URI
import java.net.URL
import java.net.URLConnection
import java.net.URLStreamHandler
import java.nio.file.Paths

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import play.utils.Resources

import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters._

class ResourcesTest {
  @Test
  def treatsGraalVmResourceUrlAsFileWhenOnlyNonSlashLookupResolves(): Unit = {
    val resourcePath: String = "/play-resource-entry"
    val classLoader: ResourcesTest.SingleResourceClassLoader = new ResourcesTest.SingleResourceClassLoader(resourcePath)
    val url: URL = ResourcesTest.resourceUrl(resourcePath)

    val isDirectory: Boolean = Resources.isDirectory(classLoader, url)

    assertThat(isDirectory).isFalse
    assertThat(classLoader.requestedResources).containsExactly(resourcePath, resourcePath + "/")
  }
}

object ResourcesTest {
  private val ResolvedResourceUrl: URL = Paths.get(".").toUri.toURL

  private def resourceUrl(path: String): URL = {
    URL.of(URI.create(s"resource:$path"), NoOpUrlStreamHandler)
  }

  final class SingleResourceClassLoader(resourcePath: String) extends ClassLoader(null) {
    private val requests: ListBuffer[String] = ListBuffer.empty

    override def getResource(name: String): URL = {
      requests += name
      if (name == resourcePath) ResolvedResourceUrl else null
    }

    def requestedResources: java.util.List[String] = requests.toSeq.asJava
  }

  private object NoOpUrlStreamHandler extends URLStreamHandler {
    override protected def openConnection(url: URL): URLConnection = {
      throw new UnsupportedOperationException("This test only inspects URL metadata")
    }
  }
}
